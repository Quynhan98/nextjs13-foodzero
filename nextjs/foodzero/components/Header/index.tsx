'use client'

import { useMemo } from 'react'
import { useRouter } from 'next/navigation'
import Link from 'next/link'
import Image from 'next/image'
import { HamburgerIcon } from '@chakra-ui/icons'
import {
  Box,
  Button,
  Container,
  Flex,
  Heading,
  IconButton,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  Text,
  useDisclosure,
} from '@chakra-ui/react'

// Constants
import { CONTACT, NAV_LIST } from '@constants/index'

// Components
import Navigation from '@components/Navigation'
import Contact from '@components/Contact'

const Header = () => {
  const router = useRouter()
  const { isOpen, onOpen, onClose } = useDisclosure()

  const handClickReservationBtn = () => {
    router.push('#reservationSection')
  }

  const renderModalMenu = useMemo(() => {
    return (
      <Modal
        isOpen={isOpen}
        onClose={onClose}
        size="full"
        data-testid="menuModal"
      >
        <ModalContent
          backgroundImage="/images/backgroundNavigation.webp"
          backgroundSize="cover"
          objectFit="cover"
          maxW="1920px"
          maxH="1080px"
          padding={{
            base: '50px 20px',
            md: '150px',
            '2xl': '222px 186px 20px 300px',
          }}
          position="relative"
          _before={{
            content: '""',
            position: 'absolute',
            width: '100%',
            height: '100%',
            top: '0px',
            left: '0px',
            backgroundColor: 'zinnwalditeBrown',
            opacity: '0.8',
            zIndex: -1,
          }}
        >
          <ModalCloseButton
            position="absolute"
            left={{ base: '15px', md: '50px' }}
            top={{ base: '15px', md: '40px' }}
            color="silverFoil"
            fontSize={{ base: '20px', '2xl': '30px' }}
          />
          <ModalBody
            display="flex"
            justifyContent={{ base: 'flex-start', md: 'space-between' }}
            flexDirection={{ base: 'column', md: 'row' }}
            gap="60px"
          >
            <Navigation onClick={onClose} navList={NAV_LIST} />
            <Flex
              width="100%"
              justifyContent="flex-end"
              pt={{ base: '0px', md: '120px', '2xl': '335px' }}
            >
              <Contact {...CONTACT} isShowIcon />
            </Flex>
          </ModalBody>
        </ModalContent>
      </Modal>
    )
  }, [isOpen])

  return (
    <Container
      as="header"
      display="flex"
      justifyContent="space-between"
      alignItems="center"
      maxW="1920px"
      padding={{
        base: '12px',
        md: '30px 50px',
        '2xl': '46px 138px 149px 118px',
      }}
      top="0px"
      left="50%"
      transform="translateX(-50%)"
      position="absolute"
      zIndex={1}
    >
      <Flex gap={{ base: '20px', md: '45px' }} alignItems="center">
        <Heading as="h1">
          <Link href="/" prefetch={false}>
            <Box
              width={{ base: '134px', lg: '307px' }}
              height={{ base: '50px', lg: '115px' }}
              position="relative"
            >
              <Image
                src="/images/logo.webp"
                alt="Logo FoodZero"
                priority
                fill
                sizes="(max-width: 768px) 307px, 115px
                (min-width: 768px) 134px, 50px
                (min-width: 1200px) 307px, 115px"
              />
            </Box>
          </Link>
        </Heading>
        <IconButton
          data-testid="buttonMenu"
          onClick={onOpen}
          variant="light"
          size="small"
          icon={
            <HamburgerIcon
              boxSize={{ base: '20px', md: '30px', '2xl': '46px' }}
            />
          }
          aria-label="menu"
        />
      </Flex>
      <Flex gap="36px" alignItems="center">
        <Box display={{ base: 'none', md: 'block' }}>
          <Link href={`tel:${CONTACT.phoneNumber}`}>
            <Text size="large" variant="primary">
              {CONTACT.phoneNumber}
            </Text>
          </Link>
        </Box>
        <Button
          onClick={handClickReservationBtn}
          variant="primary"
          fontWeight={{ base: 'normal', md: 'bold' }}
        >
          Reservations
        </Button>
      </Flex>
      {isOpen && renderModalMenu}
    </Container>
  )
}

export default Header
