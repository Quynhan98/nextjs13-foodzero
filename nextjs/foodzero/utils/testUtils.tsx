import { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'

// Themes
import { customTheme } from '@themes/index'

// Contexts
import { AuthProvider } from '@contexts/AuthProvider'
import { BookingProvider } from '@contexts/BookingProvider'

export const customRender = (component: ReactNode) => {
  return render(
    <ChakraProvider theme={customTheme}>
      <AuthProvider>
        <BookingProvider>{component}</BookingProvider>
      </AuthProvider>
    </ChakraProvider>,
  )
}

export * from '@testing-library/react'
export * from '@testing-library/jest-dom'
export { customRender as render }
