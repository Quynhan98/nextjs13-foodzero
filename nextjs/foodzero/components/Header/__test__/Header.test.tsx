// Utils
import { render } from '@utils/testUtils'

// Components
import Header from '@components/Header'

// Mocks useRouter
const useRouter = jest.spyOn(require('next/navigation'), 'useRouter')

export function mockNextUseRouter(props: {
  route: string
  pathname: string
  query: string
  asPath: string
}) {
  useRouter.mockImplementationOnce(() => ({
    route: props.route,
    pathname: props.pathname,
    query: props.query,
    asPath: props.asPath,
  }))
}

describe('Header render', () => {
  // Mocks Next.js route
  mockNextUseRouter({
    route: '/',
    pathname: '/',
    query: '',
    asPath: `/?error=${encodeURIComponent('Uh oh - something went wrong')}`,
  })

  it('Should Header match Snapshot', () => {
    const { container } = render(<Header />)

    expect(container).toMatchSnapshot()
  })
})
