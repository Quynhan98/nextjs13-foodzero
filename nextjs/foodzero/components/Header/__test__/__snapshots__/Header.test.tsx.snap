// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render Should Header match Snapshot 1`] = `
<div>
  <header
    class="chakra-container css-2mqkvj"
  >
    <div
      class="css-1psvhc8"
    >
      <h1
        class="chakra-heading css-vqcpt4"
      >
        <a
          href="/"
        >
          <img
            alt="Logo FoodZero"
            data-nimg="1"
            decoding="async"
            height="115"
            src="/_next/image?url=%2Fimages%2Flogo.webp&w=640&q=75"
            srcset="/_next/image?url=%2Fimages%2Flogo.webp&w=384&q=75 1x, /_next/image?url=%2Fimages%2Flogo.webp&w=640&q=75 2x"
            style="color: transparent;"
            width="307"
          />
        </a>
      </h1>
      <button
        aria-label="menu"
        class="chakra-button css-toj1ua"
        data-testid="buttonMenu"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-189f5bz"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="css-9b14ac"
    >
      <div
        class="css-jl5ft5"
      >
        <a
          href="tel:+86 852 346 000"
        >
          <p
            class="chakra-text css-1efn3iv"
          >
            +86 852 346 000
          </p>
        </a>
      </div>
      <button
        class="chakra-button css-ko3vlm"
        type="button"
      >
        Reservations
      </button>
    </div>
  </header>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
